# NS5004

<table>
<tr>
  <td>CheckId</td>
  <td>NS5004</td>
</tr>
<tr>
  <td>Category</td>
  <td>Usage</td>
</tr>
</table>

## Cause

Argument matcher used with WithAnyArgs. This matcher may not be executed.

## Rule description

A violation of this rule occurs when `ReturnsForAnyArgs`,`ReturnsNullForAnyArgs`,`ThrowsForAnyArgs`,`ThrowsAsyncForAnyArgs`, `WhenForAnyArgs`, `DidNotReceiveWithAnyArgs` or `ReceivedWithAnyArgs` is used in combination with arg matchers other than `Arg.Any`.

## How to fix violations

To fix a violation of this rule, replace arg matchers used in aforementioned methods with `Arg.Any`

For example:

````c#
// Incorrect:
sub.DidNotReceiveWithAnyArgs().Bar(Arg.Is(1));

// Correct:
sub.DidNotReceiveWithAnyArgs().Bar(Arg.Any<int>());

// The following are also correct. The exact arguments will not be checked due to *WithAnyArgs.
sub.DidNotReceiveWithAnyArgs().Bar(default(int));
sub.DidNotReceiveWithAnyArgs().Bar(123);
````

## How to suppress violations

This warning can be suppressed by disabling the warning in the **ruleset** file for the project.
The warning can also be suppressed programmatically for an assembly:
````c#
[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Usage", "NS5004:Argument matcher used with WithAnyArgs. This matcher may not be executed..", Justification = "Reviewed")]
````

Or for a specific code block:
````c#
#pragma warning disable NS5004 // Argument matcher used with WithAnyArgs. This matcher may not be executed..
// the code which produces warning
#pragma warning restore NS5004 // Argument matcher used with WithAnyArgs. This matcher may not be executed..
